src/lib/attach-action-provider.ts

```diff
@@ -0,0 +1,56 @@
+import { Action, StateContext } from '@ngxs/store';
+import { ActionHandlerMetaData } from '@ngxs/store/src/actions/symbols';
+import { MetaDataModel } from '@ngxs/store/src/internal/internals';
+import * as lodash from 'lodash';
+
+const META_KEY = 'NGXS_META';
+const META_OPTIONS_KEY = 'NGXS_OPTIONS_META';
+
+export function attachActionProvider(storeClass: any, providers: any[]): void {
+  if (!storeClass[META_OPTIONS_KEY]) {
+    throw new Error('storeClass is not a valid NGXS Store');
+  }
+  providers.forEach(provider => {
+    const providedActions: ActionHandlerMetaData[] = parseProvidedActions(provider);
+    providedActions.forEach(providedAction => {
+      attachToStoreClass(storeClass, provider, providedAction);
+    });
+  });
+}
+
+const parseProvidedActions = (provider: any): ActionHandlerMetaData[] => {
+  const metaDataModel: MetaDataModel = provider[META_KEY];
+  if (!metaDataModel) {
+    throw new Error('provider is not a valid NGXS Action Provider');
+  }
+  const actions: ActionHandlerMetaData[][] = Object.values(metaDataModel.actions);
+  return lodash.flattenDeep(actions);
+};
+
+const attachToStoreClass = (
+  storeClass: any,
+  provider: any,
+  action: ActionHandlerMetaData
+): void => {
+  Action({ type: action.type }, action.options)({ constructor: storeClass }, action.fn.toString());
+  const delegateDescriptor = createDelegateDescriptor(provider, action);
+  Object.defineProperty(storeClass.prototype, action.fn, delegateDescriptor);
+};
+
+const createDelegateDescriptor = (
+  provider: any,
+  action: ActionHandlerMetaData
+): PropertyDescriptor => {
+  const providerPropertyDescriptor: PropertyDescriptor = Object.getOwnPropertyDescriptor(
+    provider.prototype,
+    action.fn
+  ) as PropertyDescriptor;
+
+  const delegate = (stateContext: StateContext<any>, actionParam: any) =>
+    providerPropertyDescriptor.value.apply(provider.prototype, [stateContext, actionParam]);
+
```

Comment:
I think we need to clarify what happens when `attachActionProvider` and `attachAction` provide handlers for the the same `ActionType`. For example an `ActionA` could be provided by `attachAction` and within the ActionProvider of `attachActionProvider`.

Reply:
Make sense, do you have an idea how we can implement it? I think it is possible to check the information stored under `META_KEY` on the store class if an action is already present before adding. This is a more general inspection, not explicitly to avoid duplication caused by the redundant use of `attachAction` and `attachActionProvider`. There could also be a duplicate action which is also provided within a regular store class.
What is the error handling in `ngxs/store` when an action is provided twice in the same store class?

Reply:
@duffleit Friendly Reminder :)

Reply:
update?

Reply:
@duffleit